### PR TITLE
Prioritise and reduce the amount of events decrypted on application startup 

### PR DESCRIPTION
**Pull request by gsouquet**
View matrix-org/matrix-js-sdk#1684

### CODE REVIEW

File: src/stores/BreadcrumbsStore.ts [hunk 1]
```diff
@@ -87,6 +88,23 @@ export class BreadcrumbsStore extends AsyncStoreWithClient<IState> {
 
         this.matrixClient.on("Room.myMembership", this.onMyMembership);
         this.matrixClient.on("Room", this.onRoom);
+
+        const client = MatrixClientPeg.get();
+        const breadcrumbs = client.store.getAccountData("im.vector.setting.breadcrumbs");
```
Comment (jryans) [inline src/stores/BreadcrumbsStore.ts:93]: Hmm, it feels a bit strange to go directly to the account data here when the main purpose of this store is reading and tracking the breadcrumbs anyway... 

Maybe it fit a bit better in `updateRooms`, and you could check for the case of transitioning from no rooms to some rooms, so it only fires once...?

**Comment by jryans** [comment]
As discussed, I expect we'll need to tweak the event indexer as well, but anyway you're already looking into that. 😄

File: src/components/structures/TimelinePanel.js [hunk 2]
```diff
@@ -1141,6 +1141,19 @@ class TimelinePanel extends React.Component {
     // get the list of events from the timeline window and the pending event list
     _getEvents() {
         const events = this._timelineWindow.getEvents();
+
```
[the remark this answers is not in the document]
Comment (gsouquet) [reply src/components/structures/TimelinePanel.js:1144]: Removed, and also tracked it here https://github.com/vector-im/element-web/issues/17232

File: src/components/structures/TimelinePanel.js [hunk 3]
```diff
@@ -1141,6 +1141,18 @@ class TimelinePanel extends React.Component {
...
+        // `slice` performs a shallow copy of the array
```
Comment (turt2live) [inline src/components/structures/TimelinePanel.js:1145]: `arrayFastClone` when you're in this layer please, so we can always keep the most performant option readily available

File: src/indexing/EventIndex.js [hunk 4]
```diff
@@ -188,16 +187,11 @@ export default class EventIndex extends EventEmitter {
...
-        // If the event is not yet decrypted mark it for the
-        // Event.decrypted callback.
         if (ev.isBeingDecrypted()) {
-            const eventId = ev.getId();
-            this.liveEventsForIndex.add(eventId);
-        } else {
-            // If the event is decrypted or is unencrypted add it to the
-            // index now.
-            await this.addLiveEventToIndex(ev);
+            await ev._decryptionPromise;
```
Comment (turt2live) [inline src/indexing/EventIndex.js:191]: should get an `XXX: Private member access` to acknowledge the danger

File: src/indexing/EventIndex.js [hunk 5]
```diff
@@ -523,18 +514,18 @@ export default class EventIndex extends EventEmitter {
...
+            const decryptionPromises = matrixEvents
+                .filter(event => event.isEncrypted())
+                .map(event => {
+                    if (event.shouldAttemptDecryption()) {
+                        return event.attemptDecryption(client._crypto, {
+                            isRetry: true,
+                            emit: false,
+                        });
+                    } else {
+                        return event._decryptionPromise;
```
Comment (turt2live) [inline src/indexing/EventIndex.js:526]: maintain the private member access warning, or expose properly

File: src/stores/BreadcrumbsStore.ts [hunk 6]
```diff
@@ -22,6 +22,7 @@ import defaultDispatcher from "../dispatcher/dispatcher";
 import { arrayHasDiff } from "../utils/arrays";
 import { isNullOrUndefined } from "matrix-js-sdk/src/utils";
 import { SettingLevel } from "../settings/SettingLevel";
+import {MatrixClientPeg} from '../MatrixClientPeg';
```
Comment (turt2live) [inline src/stores/BreadcrumbsStore.ts:25]: match brace styles please:
```suggestion
import { MatrixClientPeg } from '../MatrixClientPeg';
```

File: src/stores/BreadcrumbsStore.ts [hunk 7]
```diff
@@ -59,6 +60,33 @@ export class BreadcrumbsStore extends AsyncStoreWithClient<IState> {
...
+            /**
+             * Rooms in the breadcrumb have a good chance to be interacted with
+             * again by a user. Decrypting the messages ahead of time will help
+             * reduce content shift on first render
+             */
+            this.state.rooms?.forEach(async room => {
+                const [cryptoEvent] = room.currentState.getStateEvents("m.room.encryption");
+                if (cryptoEvent) {
+                    if (!client.isRoomEncrypted(room.roomId)) {
+                        await client._crypto.onCryptoEvent(cryptoEvent);
```
Comment (turt2live) [inline src/stores/BreadcrumbsStore.ts:82]: I thought we exposed this properly? If not, should either be exposed or acknowledged as private member access

**Review by turt2live — CHANGES_REQUESTED** [review]
I am mostly being a nitpick linter here: the code looks fine, just a bunch of small things about the style.

**Review by turt2live — APPROVED** [review]
thank you for your patience with the human linter :)

**Pull request merged**